main.tf

```
@@ -205,30 +218,63 @@ data "aws_ami" "runner" {
-  instance_type        = var.instance_type
-  ebs_optimized        = var.runner_instance_ebs_optimized
-  enable_monitoring    = var.runner_instance_enable_monitoring
-  spot_price           = var.runner_instance_spot_price
-  iam_instance_profile = aws_iam_instance_profile.instance.name
-  dynamic "root_block_device" {
+resource "aws_launch_template" "gitlab_runner_instance" {
+  name_prefix            = var.runners_name
+  key_name               = var.ssh_key_pair
+  image_id               = data.aws_ami.runner.id
+  user_data              = local.template_user_data
+  instance_type          = var.instance_type
+  update_default_version = true
+  ebs_optimized          = var.runner_instance_ebs_optimized
+  monitoring {
+    enabled = var.runner_instance_enable_monitoring
+  }
+  dynamic "instance_market_options" {
+    for_each = var.runner_instance_spot_price != null && length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
+    content {
+      market_type = instance_market_options.value
+      spot_options {
+        max_price = var.runner_instance_spot_price
+      }
+    }
+  }
+  iam_instance_profile {
+    name = aws_iam_instance_profile.instance.name
+  }
+  dynamic "block_device_mappings" {
     for_each = [var.runner_root_block_device]
     content {
-      delete_on_termination = lookup(root_block_device.value, "delete_on_termination", true)
-      volume_type           = lookup(root_block_device.value, "volume_type", "gp3")
-      volume_size           = lookup(root_block_device.value, "volume_size", 8)
-      encrypted             = lookup(root_block_device.value, "encrypted", true)
-      iops                  = lookup(root_block_device.value, "iops", null)
+      device_name = lookup(block_device_mappings.value, "device_name", "/dev/xvda")
+      ebs {
+        delete_on_termination = lookup(block_device_mappings.value, "delete_on_termination", true)
+        volume_type           = lookup(block_device_mappings.value, "volume_type", "gp3")
+        volume_size           = lookup(block_device_mappings.value, "volume_size", 8)
+        encrypted             = lookup(block_device_mappings.value, "encrypted", true)
+        iops                  = lookup(block_device_mappings.value, "iops", null)
+        kms_key_id            = lookup(block_device_mappings.value, "`kms_key_id`", null)
+      }
     }
   }
-
-  associate_public_ip_address = false == var.runners_use_private_address
-
+  network_interfaces {
+    security_groups             = [aws_security_group.runner.id]
+    associate_public_ip_address = false == var.runners_use_private_address
+  }
+  tag_specifications {
+    resource_type = "instance"
+    tags          = local.tags
+  }
+  tag_specifications {
+    resource_type = "volume"
+    tags          = local.tags
+  }
+  dynamic "tag_specifications" {
+    for_each = var.runner_instance_spot_price != null && length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
```

```suggestion
    for_each = var.runner_instance_spot_price == null ? [] : ["spot"]
```

---

I would prefer to keep the test with the empty string.

Should work : 
```suggestion
    for_each = var.runner_instance_spot_price == null || var.runner_instance_spot_price == "" ? ["spot"] : []
```